File: android/app/src/main/java/com/lumoralabs/macro/data/LocalGroupStorage.kt
package com.lumoralabs.macro.data

import android.content.Context
import android.content.SharedPreferences
import com.lumoralabs.macro.domain.Group
import org.json.JSONArray
import org.json.JSONObject

object LocalGroupStorage {
    private const val PREFS_NAME = "group_storage"
    private const val GROUPS_KEY = "groups"

    fun saveGroups(context: Context, groups: List<Group>) {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val jsonArray = JSONArray()
        groups.forEach { group ->
            val obj = JSONObject()
            obj.put("id", group.id)
            obj.put("name", group.name)
            obj.put("members", JSONArray(group.members))
            jsonArray.put(obj)
        }
        prefs.edit().putString(GROUPS_KEY, jsonArray.toString()).apply()
    }

    fun loadGroups(context: Context): List<Group> {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val json = prefs.getString(GROUPS_KEY, "[]") ?: "[]"
        val jsonArray = JSONArray(json)
        val groups = mutableListOf<Group>()
        for (i in 0 until jsonArray.length()) {
            val obj = jsonArray.getJSONObject(i)
            val id = obj.getString("id")
            val name = obj.getString("name")
            val membersJson = obj.getJSONArray("members")
            val members = List(membersJson.length()) { idx -> membersJson.getString(idx) }
            groups.add(Group(id, name, members))
        }
        return groups
    }
}

File: ios/Data/LocalGroupStorage.swift
import Foundation

class LocalGroupStorage {
    private let key = "groups"
    private let defaults = UserDefaults.standard

    func saveGroups(_ groups: [Group]) {
        let array = groups.map { ["id": $0.id, "name": $0.name, "members": $0.members] }
        defaults.set(array, forKey: key)
    }

    func loadGroups() -> [Group] {
        guard let array = defaults.array(forKey: key) as? [[String: Any]] else { return [] }
        return array.compactMap { dict in
            guard let id = dict["id"] as? String,
                  let name = dict["name"] as? String,
                  let members = dict["members"] as? [String] else { return nil }
            return Group(id: id, name: name, members: members)
        }
    }
}
